### PR TITLE
Re-review test/[ command

### PR DESCRIPTION
**Pull request by AlexandreYang**
## Summary
- Re-adds the `test` and `[` builtin command implementation and all associated tests

## Test plan
- [x] Verify all `test`/`[` scenario tests pass
- [x] Run bash comparison tests

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by AlexandreYang** [comment]
**Iteration 1 self-review result:** COMMENT — 3 findings (3×P3). No P0/P1/P2 issues. Overall assessment: safe to merge.

File: tests/scenarios/cmd/test/errors/unary_option.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Derived from uutils test_test.rs — -o as unary option test
+description: "test -o treats unknown option name as false"
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply tests/scenarios/cmd/test/errors/unary_option.yaml]: Done — renamed `unary_expected.yaml` to `unary_option.yaml` to accurately reflect that the file tests `test -o arg` (unary option test returning false), not a "unary operator expected" error.

File: interp/builtins/testcmd/testcmd.go [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+			return p.parseUnaryStringOp()
+		}
+		// -o as unary tests whether a shell option is set.
+		// This restricted shell has no options, so always false.
+		if cur == "-o" {
+			p.advance() // consume -o
+			p.advance() // consume operand
+			return false
+		}
+	}
```
Comment (AlexandreYang) [inline interp/builtins/testcmd/testcmd.go]: <sub><sub>P1 Badge</sub></sub> **Missing `-a` as unary file existence test (bash compat)**

Bash treats `test -a FILE` (2 args) as a deprecated file existence test equivalent to `test -e FILE`. The current implementation treats `-a` as a bare non-empty string and then errors on the extra argument.

With `test -a /etc/passwd`:
- Bash: exit 0 (file exists)
- rshell: exit 2 (extra argument error)

With `test -a /nonexistent`:
- Bash: exit 1 (file doesn't exist)
- rshell: exit 2 (extra argument error)

Proof:
```bash
$ docker run --rm debian:bookworm-slim bash -c 'test -a /tmp; echo $?; test -a /nonexistent; echo $?'
0
1
```

**Important nuance:** `-a` is ONLY the unary file test when `remaining == 2`. With 3+ tokens, `-a` is the binary AND operator handled by `parseAnd`. For example, `test -a -a -a` returns 0 in both bash and rshell (binary AND of two non-empty strings).

**Remediation:** Add a check for `-a` as unary file test specifically when `remaining == 2`:

```suggestion
	if remaining >= 2 {
		if isUnaryFileOp(cur) {
			return p.parseUnaryFileOp()
		}
		// -a as unary file existence test (deprecated POSIX, still supported by bash).
		// Only applies with exactly 2 remaining tokens; with 3+ tokens -a is the
		// binary AND operator handled by parseAnd.
		if cur == "-a" && remaining == 2 {
			return p.parseUnaryFileOp()
		}
```

Also update `evalFileInfo` to handle `-a`:
```go
case "-e", "-a":
    return true
```

And add a scenario test that validates against bash:
```yaml
description: "test -a FILE tests file existence (deprecated POSIX)"
input:
  script: |+
    test -a /
    echo "exists: $?"
    test -a /nonexistent_path_xyz
    echo "missing: $?"
expect:
  stdout: |
    exists: 0
    missing: 1
  stderr: ""
  exit_code: 0
```

File: interp/builtins/testcmd/testcmd.go [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+		return 0, false
+	}
+	n, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		// Check for overflow — clamp to boundaries like GNU test.
+		if numErr, ok := err.(*strconv.NumError); ok && numErr.Err == strconv.ErrRange {
+			if s[0] == '-' {
+				n = math.MinInt64
+			} else {
+				n = math.MaxInt64
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/testcmd/testcmd.go]: [Claude Opus 4.6] Done — removed the clamping logic. Overflow now falls through to the "integer expression expected" error path, matching bash behavior. Removed unused `math` import and `math.MinInt64` from the symbol allowlist. Scenario test updated to match bash (removed `skip_assert_against_bash`).

File: interp/builtins/testcmd/testcmd.go [hunk 4]
```diff
@@ -0,0 +1,569 @@
...
+			return p.parseBinaryExpr()
+		}
+		// POSIX 3-arg rule: when exactly 3 tokens remain and the middle
+		// one is -a/-o, treat as binary AND/OR with string operands.
+		// e.g., "test -f -a -d" → "-f" (non-empty) AND "-d" (non-empty).
+		// Only when exactly 3 remain — with more tokens, -a/-o are handled
+		// by parseAnd/parseOr at their proper precedence level.
+		if remaining == 3 && (op == "-a" || op == "-o") {
+			return p.parseBinaryExpr()
+		}
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/testcmd/testcmd.go]: [Claude Opus 4.6] Done — changed all three occurrences of `remaining == 3` to `len(p.args) == 3` so the POSIX 3-arg disambiguation rule only fires when the total argument count to `test` is 3, not when 3 tokens remain at the current parse position inside recursive calls. Also extracted `isThreeArgBinary` and `isBinaryOpOrLogical` helpers. Added scenario test `logical/precedence.yaml` validated against bash. `test "" -a "" -o " "` now correctly returns 0.

**Comment by AlexandreYang** [comment]
**Iteration 3 self-review result**: APPROVE

**Findings**: 0 blocking (2×P3 cosmetic, same as iteration 2)

**Summary**: All prior fixes verified correct. Implementation has strong POSIX compliance, defense-in-depth safety, and comprehensive test coverage. All tests pass.

**Comment by AlexandreYang** [comment]
**Iteration 4 self-review result**: APPROVE — 0 findings. Code is clean and ready to merge.

**Comment by AlexandreYang** [comment]
@datadog @codex review

File: interp/builtins/testcmd/testcmd.go [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+// parseNot handles ! EXPR. When ! is the last remaining token, it is
+// treated as a non-empty string per POSIX single-argument rules.
+// When ! is followed by a binary operator (e.g., "! = !"), it is treated
+// as a literal string operand, not negation.
+func (p *parser) parseNot() bool {
+	if p.pos < len(p.args) && p.peek() == "!" {
+		remaining := len(p.args) - p.pos
+		if remaining == 1 {
+			p.advance()
+			return true
```
[the remark this answers is not in the document]
Comment (AlexandreYang) [reply interp/builtins/testcmd/testcmd.go]: [Claude Opus 4.6] Done — fixed in commit a563049. Changed `parseNot()` to use subexpression bounds (`p.subexprEnd - p.subexprStart == 1`) instead of global `remaining == 1` for the lone-`!` fast-path. Now `test -n x -a !` correctly returns exit 2 ("missing argument") like bash, while `test !` (1-arg) and `test x -a !` (3-arg rule) still work correctly. Added `dangling_not.yaml` scenario test validated against bash.

**Comment by AlexandreYang** [comment]
**Iteration 2 address-pr-comments result:** All 2 findings addressed and resolved.

- **P2** (`stderr_contains` will fail bash comparison): Added `skip_assert_against_bash: true`. Verified rshell already outputs "integer expression expected" matching bash — no shell code change needed. The only difference is bash's `bash: line N:` prefix.
- **P3** (prefer exact `stderr`): Replaced `stderr_contains` with exact `stderr` assertion.

All tests pass: scenario tests, bash comparison tests (`RSHELL_BASH_TEST=1`), and `gofmt` is clean.

**Comment by AlexandreYang** [comment]
**Iteration 3 self-review result: APPROVE**
- No P0/P1 findings
- 3 P2 non-blocking observations (clarifying comment suggestion, correct skip_assert usage, stderr_contains vs exact stderr convention)
- 1 P3 nit (correct behavior observation)
- All previous iteration fixes verified
- All CI passing

**Review by AlexandreYang — COMMENTED** [review]
## Code Review — Iteration 5 (Verification of iter 4 fixes)

### Verdict: **APPROVE**

### Summary

Verified the iter 4 fixes for (1) 4-arg paren disambiguation and (2) lone `(` in compound expressions. No regressions found. All tests pass, code is clean.

### Verification performed

- **Unit tests**: `go test ./interp/builtins/testcmd/...` — PASS
- **Scenario tests**: `go test ./tests/ -run TestShellScenarios` — PASS
- **gofmt**: Clean (no output from `gofmt -l`)
- **go vet**: Clean
- **Manual bash comparison**: Verified `paren_four_arg`, `dangling_not`, `overflow`, `binary_disambiguation` scenarios all match bash output

### Findings

No new issues found. The implementation is solid.

| # | Severity | Description | Status |
|---|----------|-------------|--------|
| — | — | No findings | — |

### Prior iterations summary
- Iter 1: 2x P3 fixed
- Iter 2: 1x P2 + 1x P3 fixed
- Iter 3: APPROVE
- Iter 4: 2x P2 from codex (4-arg paren, lone ( in compound) — fixed
- **Iter 5: APPROVE — fixes verified, no regressions**

**Review by AlexandreYang — COMMENTED** [review]
## Review Summary (Iteration 7) — APPROVE

**Reviewed:** `test`/`[` builtin implementation — POSIX 3/4-arg disambiguation, subexpression boundary tracking, `-a FILE` support, `==` operator, integer overflow rejection, `/dev/null` sandbox stat bypass, and all associated scenario + Go tests.

**Overall assessment: safe to merge** :white_check_mark:

| # | Priority | File | Finding |
|---|----------|------|---------|
| — | — | — | No findings |

No code changes since iteration 4. This is a re-verification confirming all prior approvals (iterations 3–6).

### Security
- The `/dev/null` stat bypass in `allowed_paths.go` is correctly scoped via `isDevNull()` — only matches the platform null device.
- No new unsafe imports (`os/exec`, `net/http`, `unsafe`). No direct filesystem access outside the sandbox.
- `findMatchingParen` has no unbounded loops (bounded by `len(args)`).
- `maxParenDepth` (128) prevents stack exhaustion from deeply nested expressions.

### Bash Compatibility
- Integer overflow now rejects with exit 2 + error message, matching bash behavior (previously clamped to MinInt64/MaxInt64).
- POSIX 3-arg and 4-arg disambiguation rules correctly handle ambiguous tokens (`!`, `(`, `-a`, `-o`) as both operators and literal strings.
- Error message changed from "extra argument" to "too many arguments" to match bash.

### Correctness
- `subexprStart`/`subexprEnd` boundary tracking is saved and restored consistently across all recursive calls (parseNot, parsePrimary with parens, 4-arg rule).
- All edge cases (lone `!`, lone `(`, `! ! = !`, `( ! ) )`, `-f -a -d`) are handled correctly per POSIX rules.

### Test Coverage
- **Adequate.** All new code paths are covered by scenario tests (12 new/modified YAML scenarios) and Go tests (pentest + functional). Bash comparison is enabled for all scenarios except those using `allowed_paths` or differing error message prefixes.

### Positive Observations
- Excellent documentation in both code comments and doc strings explaining the POSIX disambiguation rules.
- The subexpression boundary approach is a clean design that avoids the common pitfall of using global remaining-count for disambiguation.
